File: farmtech-web/src/Entity/Promotion.php
<?php

namespace App\Entity;

use App\Repository\PromotionRepository;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Validator\Constraints as Assert;

#[ORM\Entity(repositoryClass: PromotionRepository::class)]
#[ORM\Table(name: 'promotion')]
class Promotion
{
    public const TYPE_POURCENTAGE = 'pourcentage';
    public const TYPE_MONTANT_FIXE = 'montant_fixe';

    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(name: 'idPromotion', type: 'integer')]
    private ?int $idPromotion = null;

    #[ORM\Column(name: 'nom', length: 100)]
    #[Assert\NotBlank]
    private string $nom = '';

    #[ORM\Column(name: 'description', type: 'text', nullable: true)]
    private ?string $description = null;

    #[ORM\Column(name: 'typeReduction', length: 20)]
    private string $typeReduction = self::TYPE_POURCENTAGE;

    #[ORM\Column(name: 'valeurReduction', type: 'float')]
    #[Assert\Positive]
    private float $valeurReduction = 0.0;

    #[ORM\Column(name: 'dateDebut', type: 'date_immutable')]
    private ?\DateTimeImmutable $dateDebut = null;

    #[ORM\Column(name: 'dateFin', type: 'date_immutable')]
    private ?\DateTimeImmutable $dateFin = null;

    #[ORM\Column(name: 'quantiteMin', type: 'integer', options: ['default' => 1])]
    private int $quantiteMin = 1;

    #[ORM\Column(name: 'cumulable', type: 'boolean', options: ['default' => false])]
    private bool $cumulable = false;

    #[ORM\Column(name: 'actif', type: 'boolean', options: ['default' => true])]
    private bool $actif = true;

    /**
     * @var Collection<int, Produit>
     */
    #[ORM\ManyToMany(targetEntity: Produit::class, inversedBy: 'promotions')]
    #[ORM\JoinTable(name: 'promotion_produit')]
    #[ORM\JoinColumn(name: 'promotion_id', referencedColumnName: 'idPromotion', onDelete: 'CASCADE')]
    #[ORM\InverseJoinColumn(name: 'produit_id', referencedColumnName: 'idProduit', onDelete: 'CASCADE')]
    private Collection $produits;

    public function __construct()
    {
        $this->produits = new ArrayCollection();
        $today = new \DateTimeImmutable('today');
        $this->dateDebut = $today;
        $this->dateFin = $today->modify('+7 days');
    }

    public function getIdPromotion(): ?int { return $this->idPromotion; }
    public function getNom(): string { return $this->nom; }
    public function setNom(?string $nom): self { $this->nom = trim((string)$nom); return $this; }
    public function getDescription(): ?string { return $this->description; }
    public function setDescription(?string $description): self { $this->description = $description; return $this; }
    public function getTypeReduction(): string { return $this->typeReduction; }
    public function setTypeReduction(string $typeReduction): self { $this->typeReduction = $typeReduction; return $this; }
    public function getValeurReduction(): float { return $this->valeurReduction; }
    public function setValeurReduction(float $valeurReduction): self { $this->valeurReduction = $valeurReduction; return $this; }
    public function getDateDebut(): ?\DateTimeImmutable { return $this->dateDebut; }
    public function setDateDebut(?\DateTimeImmutable $dateDebut): self { $this->dateDebut = $dateDebut; return $this; }
    public function getDateFin(): ?\DateTimeImmutable { return $this->dateFin; }
    public function setDateFin(?\DateTimeImmutable $dateFin): self { $this->dateFin = $dateFin; return $this; }
    public function getQuantiteMin(): int { return $this->quantiteMin; }
    public function setQuantiteMin(int $quantiteMin): self { $this->quantiteMin = $quantiteMin; return $this; }
    public function isCumulable(): bool { return $this->cumulable; }
    public function setCumulable(bool $cumulable): self { $this->cumulable = $cumulable; return $this; }
    public function isActif(): bool { return $this->actif; }
    public function setActif(bool $actif): self { $this->actif = $actif; return $this; }
    public function getProduits(): Collection { return $this->produits; }
    public function addProduit(Produit $produit): self
    {
        if (!$this->produits->contains($produit)) {
            $this->produits->add($produit);
            if (!$produit->getPromotions()->contains($this)) {
                $produit->getPromotions()->add($this);
            }
        }

        return $this;
    }

    public function removeProduit(Produit $produit): self
    {
        if ($this->produits->removeElement($produit)) {
            $produit->getPromotions()->removeElement($this);
        }

        return $this;
    }

    public function isCurrentlyActive(): bool
    {
        if (!$this->actif || $this->dateDebut === null || $this->dateFin === null) {
            return false;
        }
        $today = new \DateTimeImmutable('today');
        return $today >= $this->dateDebut && $today <= $this->dateFin;
    }

    public function isUpcoming(): bool
    {
        return $this->actif && $this->dateDebut !== null && new \DateTimeImmutable('today') < $this->dateDebut;
    }

    public function isExpired(): bool
    {
        return $this->dateFin !== null && new \DateTimeImmutable('today') > $this->dateFin;
    }

    public function getLabel(): string
    {
        if ($this->typeReduction === self::TYPE_POURCENTAGE) {
            return '-' . rtrim(rtrim(number_format($this->valeurReduction, 2, ',', ''), '0'), ',') . '%';
        }

        return '-' . number_format($this->valeurReduction, 2, ',', ' ') . ' EUR';
    }

    public function applyTo(float $unitPrice, int $quantity): float
    {
        if ($quantity < $this->quantiteMin) {
            return $unitPrice;
        }
        if ($this->typeReduction === self::TYPE_POURCENTAGE) {
            return max(0.0, $unitPrice * (1 - ($this->valeurReduction / 100)));
        }

        return max(0.0, $unitPrice - $this->valeurReduction);
    }
}

File: pidev-3a19/src/main/java/model/Promotion.java
package model;

import javafx.beans.property.*;
import java.time.LocalDate;

/**
 * Entité Promotion — structure identique à la table Symfony.
 *
 * Colonnes : idPromotion, nom, description, typeReduction,
 *            valeurReduction, dateDebut, dateFin, quantiteMin,
 *            cumulable, actif, idProduit (NULL = global)
 *
 * typeReduction : "pourcentage" | "montant_fixe"
 */
public class Promotion {

    // Valeurs exactes stockées en BDD (minuscules, comme Symfony)
    public static final String TYPE_POURCENTAGE  = "pourcentage";
    public static final String TYPE_MONTANT_FIXE = "montant_fixe";

    // ------------------------------------------------------------------ //
    //  Propriétés JavaFX
    // ------------------------------------------------------------------ //

    private final IntegerProperty idPromotion     = new SimpleIntegerProperty();
    private final StringProperty  nom             = new SimpleStringProperty();
    private final StringProperty  description     = new SimpleStringProperty();
    private final StringProperty  typeReduction   = new SimpleStringProperty(TYPE_POURCENTAGE);
    private final DoubleProperty  valeurReduction = new SimpleDoubleProperty();
    private final ObjectProperty<LocalDate> dateDebut = new SimpleObjectProperty<>();
    private final ObjectProperty<LocalDate> dateFin   = new SimpleObjectProperty<>();
    private final IntegerProperty quantiteMin     = new SimpleIntegerProperty(1);
    private final BooleanProperty cumulable       = new SimpleBooleanProperty(false);
    private final BooleanProperty actif           = new SimpleBooleanProperty(true);
    // NULL en BDD = 0 ici (global, tous les produits)
    private final IntegerProperty idProduit       = new SimpleIntegerProperty(0);

    // ------------------------------------------------------------------ //
    //  Constructeurs
    // ------------------------------------------------------------------ //

    public Promotion() {}

    // ------------------------------------------------------------------ //
    //  Logique métier — applyTo() miroir du service Symfony
    // ------------------------------------------------------------------ //

    /**
     * Calcule le prix unitaire après application de la promotion.
     *
     * @param basePrice  prix unitaire de base
     * @param quantity   quantité commandée
     * @return prix unitaire effectif (jamais négatif)
     */
    public double applyTo(double basePrice, int quantity) {
        if (basePrice <= 0 || quantity <= 0) return basePrice;
        if (!isActif()) return basePrice;

        return switch (getTypeReduction()) {
            case TYPE_POURCENTAGE -> {
                double pct = Math.min(100.0, Math.max(0.0, getValeurReduction()));
                yield Math.max(0.0, basePrice * (1.0 - pct / 100.0));
            }
            case TYPE_MONTANT_FIXE -> Math.max(0.0, basePrice - getValeurReduction());
            default -> basePrice;
        };
    }

    /** Vérifie si la promotion est active à la date donnée. */
    public boolean isActiveOn(LocalDate date) {
        if (date == null || !isActif()) return false;
        LocalDate debut = getDateDebut();
        LocalDate fin   = getDateFin();
        if (debut != null && date.isBefore(debut)) return false;
        if (fin   != null && date.isAfter(fin))    return false;
        return true;
    }

    /** Active aujourd'hui ? */
    public boolean isActiveNow() {
        return isActiveOn(LocalDate.now());
    }

    /**
     * Libellé lisible pour l'UI.
     * Ex. : "−20 %", "−5.00 TND"
     */
    public String getLabel() {
        if (TYPE_MONTANT_FIXE.equals(getTypeReduction())) {
            return String.format("−%.2f TND", getValeurReduction());
        }
        return String.format("−%.0f %%", getValeurReduction());
    }

    // ------------------------------------------------------------------ //
    //  Getters / Setters JavaFX
    // ------------------------------------------------------------------ //

    public int getIdPromotion()                       { return idPromotion.get(); }
    public IntegerProperty idPromotionProperty()      { return idPromotion; }
    public void setIdPromotion(int v)                 { idPromotion.set(v); }

    public String getNom()                            { return nom.get(); }
    public StringProperty nomProperty()               { return nom; }
    public void setNom(String v)                      { nom.set(v); }

    public String getDescription()                    { return description.get(); }
    public StringProperty descriptionProperty()       { return description; }
    public void setDescription(String v)              { description.set(v); }

    public String getTypeReduction()                  { return typeReduction.get(); }
    public StringProperty typeReductionProperty()     { return typeReduction; }
    public void setTypeReduction(String v)            { typeReduction.set(v != null ? v : TYPE_POURCENTAGE); }

    public double getValeurReduction()                { return valeurReduction.get(); }
    public DoubleProperty valeurReductionProperty()   { return valeurReduction; }
    public void setValeurReduction(double v)          { valeurReduction.set(v); }

    public LocalDate getDateDebut()                   { return dateDebut.get(); }
    public ObjectProperty<LocalDate> dateDebutProperty() { return dateDebut; }
    public void setDateDebut(LocalDate v)             { dateDebut.set(v); }

    public LocalDate getDateFin()                     { return dateFin.get(); }
    public ObjectProperty<LocalDate> dateFinProperty() { return dateFin; }
    public void setDateFin(LocalDate v)               { dateFin.set(v); }

    public int getQuantiteMin()                       { return quantiteMin.get(); }
    public IntegerProperty quantiteMinProperty()      { return quantiteMin; }
    public void setQuantiteMin(int v)                 { quantiteMin.set(Math.max(1, v)); }

    public boolean isCumulable()                      { return cumulable.get(); }
    public BooleanProperty cumulableProperty()        { return cumulable; }
    public void setCumulable(boolean v)               { cumulable.set(v); }

    public boolean isActif()                          { return actif.get(); }
    public BooleanProperty actifProperty()            { return actif; }
    public void setActif(boolean v)                   { actif.set(v); }

    public int getIdProduit()                         { return idProduit.get(); }
    public IntegerProperty idProduitProperty()        { return idProduit; }
    public void setIdProduit(int v)                   { idProduit.set(v); }

    @Override
    public String toString() {
        return getNom() + " (" + getLabel() + ")";
    }
}
